tests/cardio_test.cpp

```diff
@@ -0,0 +1,316 @@
+//     + [ Proficiency modifiers ]    // NOT IMPLEMENTED
+//     + [ Cardio_Accumulator ]       // Adjustment based on daily activity, starting at 1/2 BMR
+//     + 10 * [ Athletics skill ]     // Formerly swimming skill, now more versatile and valuable
+//   ) * [ Trait modifiers ]          // Good/bad cardio mutations like Languorous and Hyperactive
+//
+// For NPCs (having no cardio), the formula is simply 2 * BMR.
+//
+// Important functions:
+// - Character::get_stamina
+// - Character::set_stamina
+// - Character::get_stamina_max
+// - Character::get_cardiofit
+// - Character::get_cardio_acc
+// - Character::set_cardio_acc
+// - avatar::update_cardio_acc <-- Critical
+
+static const efftype_id effect_winded( "winded" );
+
+static const move_mode_id move_mode_run( "run" );
+
+static const skill_id skill_swimming( "swimming" );
+
+// Base BMR for default character
+static const int base_bmr = 1738;
+static const int base_cardio = base_bmr;
+// Base stamina
+// MAX_STAMINA_BASE + CARDIOFIT_STAMINA_SCALING * base_bmr == 3500 + 3*1738
+static const int base_stamina = 8714;
+
+// Ensure the configured options from game_balance.json are what the tests assume they are
+static void verify_default_cardio_options()
+{
+    const int max_stamina_base = get_option<int>( "PLAYER_MAX_STAMINA_BASE" );
+    const int cardiofit_stamina_scaling = get_option<int>( "PLAYER_CARDIOFIT_STAMINA_SCALING" );
+    REQUIRE( max_stamina_base == 3500 );
+    REQUIRE( cardiofit_stamina_scaling == 3 );
+}
+
+// Count how many turns of running it takes for character to run out of stamina or become winded
+// FIXME: This underestimates total running distance by 10-15% relative to in-game measurement.
+// (For example, a normal character can run 98 steps in-game, but only 87 with this function).
+static int running_turns( Character &they, const ter_id &terrain = t_pavement )
+{
+    map &here = get_map();
+    // Please take off your shoes when entering, and no NPCs allowed
+    REQUIRE_FALSE( they.is_wearing_shoes() );
+    REQUIRE_FALSE( they.is_npc() );
+    // You put your left foot in, you put your right foot in
+    const tripoint left = they.pos();
+    const tripoint right = left + tripoint_east;
+    // You ensure two tiles of terrain to hokey-pokey in
+    here.ter_set( left, terrain );
+    here.ter_set( right, terrain );
+    REQUIRE( here.ter( left ) == terrain );
+    REQUIRE( here.ter( right ) == terrain );
+    // Take a deep breath and start running
+    they.set_stamina( they.get_stamina_max() );
+    they.set_movement_mode( move_mode_run );
+    they.update_body();
+    // Count how many turns it takes to become winded
+    int turns = 0;
+    int STOP = 1000; // Safe exit in case of Superman
+    int last_stamina = they.get_stamina_max();
+    // Run until out of stamina or winded (should happen at the same time)
+    while( they.get_stamina() > 0 && !they.has_effect( effect_winded ) && turns < STOP ) {
```

Comment:
I strongly suspect that the reason this doesn't match up with in-game testing is that one step is not exactly equal to one turn, especially when running.
What you should be doing here instead of counting turns is decrementing u.moves and ticking over to the next turn wwhen moves is less than or equal to zero.

Reply:
Thanks, hopefully I grokked right (ef74300). I've ensured that `they.moves` are being decremented by the call to `g->walk_move`, and am now doing `they.moves += they.get_speed()` when it goes to zero or below.

Resulting numbers are a bit *lower* than before (further from in-game observations), so obviously there's something else I'm missing.

Reply:
Re-confirming in-game measurements shows they are actually quite close to test results now.